### PR TITLE
Several commits related to OCP4 CIS 1.2

### PR DESCRIPTION
**Pull request by jhrozek**
Details are spelled out in the commit messages. tl;dr adds a bunch
of small fixes and tests for rules that are mostly good. Adds several new
rules and removes rules that didn't work for OCP4.

#### Description:

- More ocp4 content

### CODE REVIEW

File: applications/openshift/api-server/api_server_admission_control_plugin_Scc/rule.yml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+title: 'Enable the SecurityContextConstraint Admission Control Plugin'
+
+description: |-
+   To ensure pod permissions are managed, make sure that the
+   <tt>SecurityContextConstraint</tt> admission control plugin is used.
+
+rationale: |-
+  SCCs to define a set of conditions that a pod must run with in order to
+  be accepted into the system.
```
Comment (redhatrises) [inline applications/openshift/api-server/api_server_admission_control_plugin_Scc/rule.yml:13]: That's nice, but what is the security rationale?

File: applications/openshift/api-server/api_server_api_priority_flowschema_catch_all/rule.yml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    To make sure that all requests are controlled by the API Priority
+    and Fairness feature, make sure that there exists a catch-all rule.
+
+rationale: |-
+    Using <tt>APIPriorityAndFairness</tt> feature provides a fine-grained way
+    to control the behaviour of the Kubernetes API server in an overload
+    situation. The well-known FlowSchema <tt>catch-all</tt> should be available
+    to make sure that every request gets some kind of classification. By default,
+    the <tt>catch-all</tt> priority level only allows one concurrency share and
+    does not queue requests
```
Comment (redhatrises) [inline applications/openshift/api-server/api_server_api_priority_flowschema_catch_all/rule.yml:17]: That's nice, but what is the security rationale?

File: applications/openshift/api-server/api_server_api_priority_flowschema_catch_all/rule.yml [hunk 3]
```diff
@@ -0,0 +1,47 @@
+documentation_complete: true
+
+prodtype: ocp4
+
+title: 'Make sure there exists a catch-all FlowSchema object for API Priority and Fairness'
```
Comment (redhatrises) [inline applications/openshift/api-server/api_server_api_priority_flowschema_catch_all/rule.yml:5]:
```suggestion
title: 'Ensure catch-all FlowSchema object for API Priority and Fairness Exists'
```

File: applications/openshift/api-server/api_server_api_priority_gate_enabled/rule.yml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+title: 'Enable the APIPriorityAndFairness feature gate'
+
+description: |-
+    To limit the rate at which the API Server accepts requests, make
+    sure that the API Priority and Fairness feature is enabled.
+
+rationale: |-
+    Using <tt>APIPriorityAndFairness</tt> feature provides a fine-grained way
+    to control the behaviour of the Kubernetes API server in an overload
+    situation.
```
Comment (redhatrises) [inline applications/openshift/api-server/api_server_api_priority_gate_enabled/rule.yml:14]: That's nice, but what is the security rationale?

File: applications/openshift/api-server/api_server_auth_mode_rbac/rule.yml [hunk 5]
```diff
@@ -0,0 +1,38 @@
+prodtype: ocp4
+
+title: Must have authorization-mode RBAC set
```
Comment (redhatrises) [inline applications/openshift/api-server/api_server_auth_mode_rbac/rule.yml:3]:
```suggestion
title: Ensure authorization-mode RBAC is configured
```

File: applications/openshift/api-server/api_server_auth_mode_rbac/rule.yml [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+description: 'Turn on Role Based Access Control'
```
Comment (redhatrises) [inline applications/openshift/api-server/api_server_auth_mode_rbac/rule.yml:5]: Description needs to be flushed out more. How is this done?

File: applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/rule.yml [hunk 7]
```diff
@@ -0,0 +1,45 @@
+documentation_complete: true
+
+prodtype: ocp4
+
+title: 'Check that no admission control plugins are disabled'
```
Comment (redhatrises) [inline applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/rule.yml:5]:
```suggestion
title: 'Ensure all admission control plugins are enabled'
```

File: applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/rule.yml [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+description: |-
+    To make sure none of them is explicitly enabled, call
+    <pre>$ oc -n openshift-kube-apiserver get configmap config -o json | jq -r '.data."config.yaml"' | jq '.apiServerArguments."disable-admission-plugins"'</pre>
```
Comment (redhatrises) [inline applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/rule.yml:9]: Are they supposed to be enabled or disabled?

File: applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/rule.yml [hunk 9]
```diff
@@ -0,0 +1,45 @@
+documentation_complete: true
+
+prodtype: ocp4
+
+title: 'Check that no admission control plugins are disabled'
+
+description: |-
+    To make sure none of them is explicitly enabled, call
```
Comment (redhatrises) [inline applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/rule.yml:8]:
```suggestion
    To make sure none of them is explicitly enabled, run the following command:
```

File: applications/openshift/api-server/api_server_auth_mode_rbac/rule.yml [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+severity: medium
+
+references:
+    cis: 1.2.9
+
+ocil_clause: 'The RBAC authorization mode is enabled'
+
+ocil: |-
+    Run the following command:
```
Comment (redhatrises) [inline applications/openshift/api-server/api_server_auth_mode_rbac/rule.yml:20]:
```suggestion
    To verify that RBAC authorization mode is enabled, run the following command:

File: applications/openshift/api-server/api_server_basic_auth/rule.yml [hunk 11]
```diff
@@ -32,10 +32,10 @@ severity: medium
 references:
     cis: 1.2.2
 
-ocil_clause: 'basic-auth-file is configured and enabled on the master node'
+ocil_clause: 'basic-auth-file is configured and enabled for the API server'
 
 ocil: |-
-    Run the following command on the master node(s):
+    Run the following command:
```
Comment (redhatrises) [inline applications/openshift/api-server/api_server_basic_auth/rule.yml:38]:
```suggestion
    To verify that <tt>basic-auth-file</tt> is configured and enabled for the API server, run the following command:
```

File: applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/rule.yml [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+    <pre>$ oc -n openshift-kube-apiserver get configmap config -o json | jq -r '.data."config.yaml"' | jq '.apiServerArguments."disable-admission-plugins"'</pre>
+
+rationale: |-
+    Several hardening controls depend on certain API server admission plugins
+    being enabled.
```
Comment (redhatrises) [inline applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/rule.yml:13]: That's great, but rationale should be flushed out more.

File: applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/rule.yml [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    cis: 1.2.14
+    cis: 1.2.14
+    cis: 1.2.15
+    cis: 1.2.16
+    cis: 1.2.17
+
+ocil_clause: 'No admission plugins are disabled'
+
+ocil: |-
+    The list of disabled admission plugins should be empty:
```
Comment (redhatrises) [inline applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/rule.yml:28]:
```suggestion
    To verify that the list of disabled admission plugins is empty, run the following command:
```

File: applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/rule.yml [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+    <pre>$ oc -n openshift-kube-apiserver get configmap config -o json | jq -r '.data."config.yaml"' | jq '.apiServerArguments."disable-admission-plugins"'</pre>
```
Comment (redhatrises) [inline applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/rule.yml:29]: There should be no output.

File: applications/openshift/api-server/api_server_auth_mode_rbac/rule.yml [hunk 15]
```diff
@@ -0,0 +1,38 @@
+prodtype: ocp4
+
+title: Must have authorization-mode RBAC set
+
+description: 'Turn on Role Based Access Control'
+
+rationale: |-
+  Role Based Access Control (RBAC) allows fine-grained control over the
+  operations that different entities can perform on different objects in
+  the cluster.
```
Comment (redhatrises) [inline applications/openshift/api-server/api_server_auth_mode_rbac/rule.yml:10]: Cool. Some more sentences on security would be great.

File: applications/openshift/api-server/api_server_api_priority_gate_enabled/rule.yml [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+
+severity: medium
+
+references:
+    cis: 1.2.10
+
+ocil_clause: '<tt>.apiServerArguments["feature-gates"]</tt> does not include <tt>EventRateLimit</tt>'
+
+ocil: |-
+    Run the following command:
```
Comment (redhatrises) [inline applications/openshift/api-server/api_server_api_priority_gate_enabled/rule.yml:24]:
```suggestion
    To verify that <tt>APIPriorityAndFairness<tt> is enabled, run the following command:
```

File: applications/openshift/api-server/api_server_api_priority_gate_enabled/rule.yml [hunk 17]
```diff
@@ -0,0 +1,44 @@
+documentation_complete: true
+
+prodtype: ocp4
+
+title: 'Enable the APIPriorityAndFairness feature gate'
+
+description: |-
+    To limit the rate at which the API Server accepts requests, make
+    sure that the API Priority and Fairness feature is enabled.
```
Comment (redhatrises) [inline applications/openshift/api-server/api_server_api_priority_gate_enabled/rule.yml:9]: Need to add how it is enabled in the description.

File: applications/openshift/api-server/api_server_admission_control_plugin_AlwaysAdmit/rule.yml [hunk 18]
```diff
@@ -8,22 +8,21 @@ description: |-
...
 references:
     cis: 1.2.11
 
-ocil_clause: '<tt>admissionConfig</tt> is not set'
+ocil_clause: '<tt>AlwaysAllow</tt> admission control plugin is not set'
 
 ocil: |-
-    Run the following command on the master node(s):
-    <pre>$ oc get configmap config -n openshift-kube-apiserver -ojson | jq -r '.data["config.yaml"]' | grep 'AlwaysAdmit'</pre>
+    Run the following command:
```
Comment (redhatrises) [inline applications/openshift/api-server/api_server_admission_control_plugin_AlwaysAdmit/rule.yml:24]: To verify.....

File: applications/openshift/api-server/api_server_admission_control_plugin_NamespaceLifecycle/rule.yml [hunk 19]
```diff
@@ -19,12 +19,12 @@ severity: medium
 references:
     cis: 1.2.15
 
-ocil_clause: '<tt>admissionConfig</tt> does not contain <tt>NamespaceLifecycle</tt>'
+ocil_clause: 'API server config contains <tt>NamespaceLifecycle</tt>'
 
 ocil: |-
-    The ServiceAccount plugin should be enabled in the list of enabled plugins in
+    The NamespaceLifecycle plugin should be enabled in the list of enabled plugins in
```
Comment (redhatrises) [inline applications/openshift/api-server/api_server_admission_control_plugin_NamespaceLifecycle/rule.yml:25]: To verify that....

File: applications/openshift/api-server/api_server_auth_mode_no_aa/rule.yml [hunk 20]
```diff
@@ -12,10 +12,10 @@ severity: medium
 references:
     cis: 1.2.7
 
-ocil_clause: 'The Node authorization-mode is configured and enabled on the master node'
+ocil_clause: 'The Node authorization-mode is configured and enabled'
 
 ocil: |-
-    Run the following command on the master node(s):
+    Run the following command:
```
Comment (redhatrises) [inline applications/openshift/api-server/api_server_auth_mode_no_aa/rule.yml:18]: To verify that....

File: applications/openshift/api-server/api_server_admission_control_plugin_Scc/rule.yml (hunk 1 above)
Comment (jhrozek) [reply applications/openshift/api-server/api_server_admission_control_plugin_Scc/rule.yml:13]: Fixed locally, will mark as resolved when I push new patches.

File: applications/openshift/api-server/api_server_api_priority_flowschema_catch_all/rule.yml (hunk 3 above)
Comment (jhrozek) [reply applications/openshift/api-server/api_server_api_priority_flowschema_catch_all/rule.yml:5]: Thanks, I used your suggestion.

File: applications/openshift/api-server/api_server_api_priority_flowschema_catch_all/rule.yml (hunk 2 above)
Comment (jhrozek) [reply applications/openshift/api-server/api_server_api_priority_flowschema_catch_all/rule.yml:17]: Fixed locally, will mark as resolved when I push new patches.

File: applications/openshift/api-server/api_server_api_priority_gate_enabled/rule.yml (hunk 4 above)
Comment (jhrozek) [reply applications/openshift/api-server/api_server_api_priority_gate_enabled/rule.yml:14]: Fixed locally, will mark as resolved when I push new patches.

File: applications/openshift/api-server/api_server_auth_mode_rbac/rule.yml (hunk 5 above)
Comment (jhrozek) [reply applications/openshift/api-server/api_server_auth_mode_rbac/rule.yml:3]: Thanks, I used your suggestion.

File: applications/openshift/api-server/api_server_auth_mode_rbac/rule.yml (hunk 6 above)
Comment (jhrozek) [reply applications/openshift/api-server/api_server_auth_mode_rbac/rule.yml:5]: Fixed locally, will mark as resolved when I push new patches.

File: applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/rule.yml (hunk 7 above)
Comment (jhrozek) [reply applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/rule.yml:5]: Thanks, I used your suggestion.

File: applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/rule.yml (hunk 8 above)
Comment (jhrozek) [reply applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/rule.yml:9]: oops :)

File: applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/rule.yml (hunk 9 above)
Comment (jhrozek) [reply applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/rule.yml:8]: thanks

File: applications/openshift/api-server/api_server_auth_mode_rbac/rule.yml (hunk 10 above)
Comment (jhrozek) [reply applications/openshift/api-server/api_server_auth_mode_rbac/rule.yml:20]: thanks

File: applications/openshift/api-server/api_server_basic_auth/rule.yml (hunk 11 above)
Comment (jhrozek) [reply applications/openshift/api-server/api_server_basic_auth/rule.yml:38]: thanks

File: applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/rule.yml (hunk 12 above)
Comment (jhrozek) [reply applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/rule.yml:13]: what about now?

File: applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/rule.yml (hunk 13 above)
Comment (jhrozek) [reply applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/rule.yml:28]: thanks

File: applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/rule.yml (hunk 14 above)
Comment (jhrozek) [reply applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/rule.yml:29]: thanks

File: applications/openshift/api-server/api_server_auth_mode_rbac/rule.yml (hunk 15 above)
Comment (jhrozek) [reply applications/openshift/api-server/api_server_auth_mode_rbac/rule.yml:10]: added something, can you check?

File: applications/openshift/api-server/api_server_api_priority_gate_enabled/rule.yml (hunk 16 above)
Comment (jhrozek) [reply applications/openshift/api-server/api_server_api_priority_gate_enabled/rule.yml:24]: thanks

File: applications/openshift/api-server/api_server_api_priority_gate_enabled/rule.yml (hunk 17 above)
Comment (jhrozek) [reply applications/openshift/api-server/api_server_api_priority_gate_enabled/rule.yml:9]: added

File: applications/openshift/api-server/api_server_admission_control_plugin_AlwaysAdmit/rule.yml (hunk 18 above)
Comment (jhrozek) [reply applications/openshift/api-server/api_server_admission_control_plugin_AlwaysAdmit/rule.yml:24]: fixed

File: applications/openshift/api-server/api_server_admission_control_plugin_NamespaceLifecycle/rule.yml (hunk 19 above)
Comment (jhrozek) [reply applications/openshift/api-server/api_server_admission_control_plugin_NamespaceLifecycle/rule.yml]: fixed

File: applications/openshift/api-server/api_server_auth_mode_no_aa/rule.yml (hunk 20 above)
Comment (jhrozek) [reply applications/openshift/api-server/api_server_auth_mode_no_aa/rule.yml]: fixed

**Comment by jhrozek** [comment]
Thank you for the review, hopefully I fixed all the issues you found.

File: applications/openshift/api-server/api_server_admission_control_plugin_Scc/rule.yml [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+   <tt>SecurityContextConstraint</tt> admission control plugin is used.
+
+rationale: |-
+    A Security Context Constraint is a cluster-level resource that controls the actions
+    which a pod can perform and what the pod may access. The
+    <tt>SecurityContextConstraint</tt> objects define a set of conditions that a pod
+    must run with in order to be accepted into the system. Security Context Constraints
+    are comprised of settings and strategies that control the security features
+    a pod has access to and hence this must be used to control pod access
+    permissions.
```
Comment (redhatrises) [inline applications/openshift/api-server/api_server_admission_control_plugin_Scc/rule.yml:18]: This is a good security rationale... good job!

File: applications/openshift/api-server/api_server_api_priority_flowschema_catch_all/rule.yml [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+
+title: 'Ensure catch-all FlowSchema object for API Priority and Fairness Exists'
+
+description: |-
+    Using <tt>APIPriorityAndFairness</tt> feature provides a fine-grained way
+    to control the behaviour of the Kubernetes API server in an overload
+    situation. The well-known FlowSchema <tt>catch-all</tt> should be available
+    to make sure that every request gets some kind of classification. By default,
+    the <tt>catch-all</tt> priority level only allows one concurrency share and
+    does not queue requests
```
Comment (redhatrises) [inline applications/openshift/api-server/api_server_api_priority_flowschema_catch_all/rule.yml:13]: Can you add the `oc get flowschema` commands here?

File: applications/openshift/api-server/api_server_admission_control_plugin_Scc/rule.yml (hunk 21 above)
Comment (redhatrises) [reply applications/openshift/api-server/api_server_admission_control_plugin_Scc/rule.yml:18]: Read the other rationale's in this PR and they are very good.

File: applications/openshift/api-server/api_server_api_priority_flowschema_catch_all/rule.yml (hunk 22 above)
Comment (jhrozek) [reply applications/openshift/api-server/api_server_api_priority_flowschema_catch_all/rule.yml]: done

**Comment by jhrozek** [comment]
Thank you again for the review, I pushed a new patchset. Only the patch titled 'CIS 1.2.10: Add a rule with disabled OVAL to check for API priority and fairness' has changed.

**Review by JAORMX — APPROVED** [review]
We'll fix the permission issues for the opreator in the next sprint.

**Pull request merged**